src/bridge/cockpitdbusmachines.c

```diff
@@ -0,0 +1,416 @@
+          json_node_free (j);
+        }
+    }
+
+  g_free (conf_glob.gl_pathv[0]); /* we allocated that ourselves */
+  globfree (&conf_glob);
+  g_free (glob_str);
+
+  return machines;
+}
+
+/* returns a floating GVariant */
+static GVariant *
+build_machines (void)
+{
+  JsonNode *machines;
+  GError *error = NULL;
+  GVariant *variant;
+
+  machines = read_machines_json ();
+  variant = json_gvariant_deserialize (machines, MACHINES_SIG, &error);
+  /* if the signature does not match, we screwed up in the parser already */
+  g_assert (variant != NULL);
+  json_node_free (machines);
+  return variant;
+}
+
+/* iterator function for update_machine() */
+static void
+update_machine_property(JsonObject *object,
+                        const gchar *member_name,
+                        JsonNode *member_node,
+                        gpointer user_data)
+{
+  json_object_set_member ((JsonObject*) user_data, member_name, json_node_copy (member_node));
+}
+
+static gboolean
+update_machine (const char *filename, const char* hostname, JsonNode *info, GError **error)
+{
+  gchar *path;
+  JsonNode *cur_config;
+  JsonGenerator *json_gen;
+  JsonObject *cur_config_obj;
+  JsonNode *cur_props;
+  gboolean res;
+
+  g_assert (JSON_NODE_HOLDS_OBJECT (info));
+
+  path = g_build_filename (machines_json_dir(), filename, NULL);
+  cur_config = parse_json_file (path);
+  if (cur_config == NULL)
+    cur_config = new_object_node ();
+  cur_config_obj = json_node_get_object (cur_config);
+  cur_props = json_object_get_member (cur_config_obj, hostname);
+
+  if (cur_props)
+    {
+      /* update settings for hostname */
+      g_assert (JSON_NODE_HOLDS_OBJECT(cur_props));
+      json_object_foreach_member (json_node_get_object (info), update_machine_property, json_node_get_object (cur_props));
+    }
+  else
+    /* create new entry for host name */
+    json_object_set_member (cur_config_obj, hostname, json_node_copy (info));
+
+  /* write back json file */
+  json_gen = json_generator_new ();
+  json_generator_set_root (json_gen, cur_config);
+  json_generator_set_pretty (json_gen, TRUE); /* bikeshed zone */
+  res = json_generator_to_file (json_gen, path, error);
+  g_free (path);
+  g_object_unref (json_gen);
+  json_node_free (cur_config);
+  return res;
+}
+
+static GVariant *
+machines_get_property (GDBusConnection *connection,
+                       const gchar *sender,
+                       const gchar *object_path,
+                       const gchar *interface_name,
+                       const gchar *property_name,
+                       GError **error,
+                       gpointer user_data)
+{
+  g_return_val_if_fail (property_name != NULL, NULL);
+
+  if (g_str_equal (property_name, "Machines"))
+    return build_machines ();
+  else
+    g_return_val_if_reached (NULL);
+}
+
+static void
+machines_method_call (GDBusConnection *connection,
```

Comment:
I thought ATM remotectl only managed the certificate, not `machines.json` entries? Adding a CLI is indeed [a work item in the Trello card](https://trello.com/c/WSEOANNY/268-finalize-machines-json-format) but hasn't been discussed/designed yet. My first version had the `99-webui.json` file name hardcoded in the bridge, but I changed it into an explicit D-Bus argument so that it only needs to be hardcoded in the frontend and thus the D-Bus API is flexible to write other files, for a CLI API to supply the "pre-seeded" files.

I think a CLI for adding a machine should encourage to specify a file name as well, maybe defaulting to "10-<hostname>.json"?